action.py

```diff
@@ -0,0 +1,276 @@
+    boto3.client("s3").put_object(
+        Bucket=application_version_bucket,
+        Body=deployment_archive.read_bytes(),
+        Key=deployment_archive.name,
+    )
+
+
+def create_and_upload_deployment_archive(
+        deployment_archive: str,
+        container_configs: list[ContainerConfig],
+        version_label: str,
+        application_version_bucket: str,
+):
+    output_path = Path(".build-artifacts")
+    output_path.mkdir(exist_ok=True)
+    output_file = output_path / deployment_archive
+
+    create_deployment_archive(
+        output_file=output_file,
+        config=container_configs,
+        version_label=version_label,
+    )
+
+    upload_deployment_archive_to_s3(application_version_bucket, output_file)
+
+
+def build_image(container_configs: ContainerConfig, dockerfile_path: Path, build_path: Path):
+    run_command(f"docker build -t {container_configs.name}-ci -f {build_path / dockerfile_path} {build_path}")
+
+
+def upload_image_to_ecr(container_configs: ContainerConfig, version_label: str):
+    image_destination = container_configs.image_url(version_label)
+    run_commands(
+        (
+            f"docker tag {container_configs.name}-ci {image_destination}",
+            "$(aws ecr get-login --no-include-email)",
+            f"docker push {image_destination}",
+        ),
+    )
+
+
+def build_and_upload_images(container_configs: list[ContainerConfig], base_path: str, version_label: str):
+    for container in container_configs:
+        build_image(container, Path(container.dockerfile_path), Path(base_path))
+        upload_image_to_ecr(container, version_label)
+
+
+def create_beanstalk_application_version(app: str, version: str, description: str, bucket: str, archive: str):
+    boto3.client("elasticbeanstalk").create_application_version(
+        ApplicationName=app,
+        VersionLabel=version,
+        Description=description,
+        SourceBundle={
+            "S3Bucket": bucket,
+            "S3Key": archive,
+        },
+        Process=True,
+    )
+
+
+def get_deployment_status(environment_name: str) -> dict[str, ...]:
+    return boto3.client("elasticbeanstalk").describe_environment_health(
+        EnvironmentName=environment_name,
+        AttributeNames=[
+            "HealthStatus",
+            "InstancesHealth",
+            "Causes",
+            "Color",
+            "Status",
+        ],
+    )
+
+
+def is_version_deployed(environment_name: str, version: str, application_name: str) -> bool:
+    return bool(boto3.client("elasticbeanstalk").describe_environments(
+        ApplicationName=application_name,
+        VersionLabel=version,
+        EnvironmentNames=[environment_name],
+    )["Environments"])
+
+
+def wait_until_deployment_successful_finished(
+        application_name: str,
+        environment_name: str,
+        version_label: str,
+        wait_timeout: Optional[int] = 20 * 60,
+        wait_time_step: Optional[int] = 8,
```

Comment:
```suggestion
        wait_timeout: int = 20 * 60,
        wait_time_step: int = 8,
```
Optional heisst nicht "hat ein default Wert" sondern "kann None sein"

Reply:
Danke, ich dachte, wenn ich einen `default` setze, ist es immer ein `Optional`.

Reply:
Das Thema haben wir schonmal besprochen - ja, aber nur wenn default `None` ist:

https://stackoverflow.com/questions/51710037/how-should-i-use-the-optional-type-hint